File: services/shared/shared-events/src/main/java/ru/rutcampustrack/shared/events/IdempotencyGuard.java
package ru.rutcampustrack.shared.events;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.UUID;

/**
 * Helper для consumer-side dedup по {@code event_id} (M13 G8).
 *
 * <p>Использование в {@code @RabbitListener}-метода:
 * <pre>{@code
 * @RabbitListener(queues = "academic-service.events")
 * @Transactional
 * public void onEvent(Map<String, Object> envelope) {
 *     if (!idempotencyGuard.tryClaim("academic", envelope)) {
 *         return; // duplicate, уже обработано
 *     }
 *     // ... основной handler ...
 * }
 * }</pre>
 *
 * <p>Поведение при отсутствии {@code event_id} в envelope (M13 G24-fix-6):
 * helper пробрасывает {@link IllegalStateException} с ERROR-логом
 * (fail-closed). После M13 G8 все наши publisher'ы заполняют event_id
 * через {@link AbstractEventPublisher#fillDefaults} — событие без id =
 * либо bug в новом publisher'е, либо forged message. Раньше (до G24-fix-6)
 * fail-open возвращал true → security-auditor отметил это как H3:
 * compromise broker creds + drop event_id = bypass replay-защиты. Теперь
 * @Transactional handler rollback → DLQ → manual triage.
 *
 * <p>Helper потокобезопасен — не хранит state, всё в {@link IdempotencyStore}.
 */
public class IdempotencyGuard {

    private static final Logger log = LoggerFactory.getLogger(IdempotencyGuard.class);

    private final IdempotencyStore store;

    public IdempotencyGuard(IdempotencyStore store) {
        this.store = store;
    }

    /**
     * Пытается claim'ить событие из envelope. Возвращает {@code true}
     * если consumer должен обработать, {@code false} если duplicate.
     *
     * <p>Если в envelope нет валидного {@code event_id} — кидает
     * {@link IllegalStateException} (M13 G24-fix-6 fail-closed). Caller
     * через @Transactional handler делает rollback, message идёт в DLQ.
     *
     * @param consumerId логический id consumer'а
     * @param envelope   сырой envelope (Map из Jackson после @RabbitListener)
     * @throws IllegalStateException если event_id отсутствует или malformed
     */
    public boolean tryClaim(String consumerId, Map<String, Object> envelope) {
        UUID eventId = extractEventId(envelope);
        if (eventId == null) {
            String eventType = envelope == null ? null : (String) envelope.get("event_type");
            log.error("Idempotency: rejecting event без event_id (fail-closed) — "
                            + "consumer={}, event_type={}. После M13 G8 все publisher'ы "
                            + "обязаны заполнять event_id (AbstractEventPublisher.fillDefaults).",
                    consumerId, eventType);
            throw new IllegalStateException(
                    "Event без event_id rejected (fail-closed, M13 G24-fix-6): "
                            + "consumer=" + consumerId + ", event_type=" + eventType);
        }
        boolean claimed = store.tryClaim(consumerId, eventId);
        if (!claimed) {
            String eventType = (String) envelope.get("event_type");
            log.info("Idempotency: duplicate event skipped (consumer={}, event_type={}, event_id={})",
                    consumerId, eventType, eventId);
        }
        return claimed;
    }

    /**
     * Извлекает {@code event_id} из envelope. Поддерживает оба представления:
     * Jackson по умолчанию десериализует UUID в String, но если caller
     * подал готовый UUID — тоже принимаем.
     */
    static UUID extractEventId(Map<String, Object> envelope) {
        if (envelope == null) {
            return null;
        }
        Object raw = envelope.get("event_id");
        if (raw == null) {
            return null;
        }
        if (raw instanceof UUID uuid) {
            return uuid;
        }
        try {
            return UUID.fromString(raw.toString());
        } catch (IllegalArgumentException ex) {
            log.warn("Idempotency: malformed event_id '{}', processing without dedup", raw);
            return null;
        }
    }
}

File: services/notification-bot/bot/services/idempotency_guard.py
"""Consumer-side dedup по event_id для notification-bot (M13 G8).

Java-side имеет shared-events `IdempotencyGuard` + `IdempotencyStore`,
тут эквивалент через Redis: ключ `consumed:{consumer_id}:{event_id}`,
TTL 1 час (события старше уже не редоставляются Rabbit'ом).

Использование:
    guard = BotIdempotencyGuard(host="redis", port=6379, password="...")
    if not await guard.try_claim(event_id):
        return  # duplicate, skip
"""

import logging
from typing import Optional

import redis.asyncio as aioredis

logger = logging.getLogger(__name__)


class BotIdempotencyGuard:
    """Redis-backed dedup для bot consumer'а (M13 G8).

    Атомарный `SET key value NX EX ttl` гарантирует что:
    - первый caller получит claim (return True);
    - все последующие в течение TTL получат False (skip duplicate).

    Если `event_id` отсутствует — пропускаем (return True) с warn'ом.
    Это совместимость с pre-M13 publisher'ами; после M13 G8 все события
    содержат `event_id` (Java `AbstractEventPublisher.fillDefaults`).
    """

    DEFAULT_CONSUMER_ID = "notification-bot"
    DEFAULT_TTL_SECONDS = 3600  # 1h — больше чем Rabbit redelivery window

    def __init__(
        self,
        consumer_id: str = DEFAULT_CONSUMER_ID,
        ttl_seconds: int = DEFAULT_TTL_SECONDS,
        host: str = "redis",
        port: int = 6379,
        password: str = "",
        redis_client: Optional[aioredis.Redis] = None,
    ) -> None:
        if redis_client is not None:
            self._redis = redis_client
        else:
            auth = f":{password}@" if password else ""
            url = f"redis://{auth}{host}:{port}"
            self._redis = aioredis.from_url(
                url, max_connections=10, decode_responses=True
            )
        self._consumer_id = consumer_id
        self._ttl = ttl_seconds

    def _key(self, event_id: str) -> str:
        return f"consumed:{self._consumer_id}:{event_id}"

    async def try_claim(self, event_id: Optional[str]) -> bool:
        """Пытается claim'ить event_id. True если новый, False если дубль.

        - None / blank event_id → ValueError (M13 G24-fix-6 fail-closed).
          После M13 G8 все publisher'ы обязаны заполнять event_id
          (Java AbstractEventPublisher.fillDefaults). Событие без id =
          либо bug в новом publisher'е, либо forged message в обход
          replay-защиты. Caller (event_consumer) делает NACK → DLQ.
        - Redis ошибка → пробрасывает исключение (M13 G24-fix-2: ранее
          fail-open возвращал True, что в паре с handler-exception swallow
          делало replay-защиту бесполезной). Теперь caller реджектит
          message → DLQ → manual triage.
        """
        if not event_id:
            logger.error(
                "Idempotency: rejecting event без event_id (fail-closed) — consumer=%s. "
                "После M13 G8 все publisher'ы обязаны заполнять event_id.",
                self._consumer_id,
            )
            raise ValueError(
                f"Event без event_id rejected (fail-closed, M13 G24-fix-6): "
                f"consumer={self._consumer_id}"
            )
        key = self._key(event_id)
        # SET key value NX EX ttl — атомарный insert-or-fail.
        # Redis exception пробрасывается — caller отлавливает и реджектит
        # message в DLQ (см. event_consumer.py).
        ok = await self._redis.set(key, "1", nx=True, ex=self._ttl)
        if ok:
            return True
        logger.info(
            "Idempotency: duplicate event skipped (consumer=%s, event_id=%s)",
            self._consumer_id,
            event_id,
        )
        return False

    async def close(self) -> None:
        await self._redis.aclose()
